[Coverage] Make additional counters available for BranchRegion. NFC.

chapuni opened the pull request:
`getBranchCounterPair()` allocates an additional Counter to SkipPath in `SingleByteCoverage`.

`IsCounterEqual()` calculates the comparison with rewinding counter replacements.

`NumRegionCounters` is updated to take additional counters in account.

`incrementProfileCounter()` has a few additiona arguments.

- `UseSkipPath=true`, to specify setting counters for SkipPath. It assumes `UseSkipPath=false` is used together.

- `UseBoth` may be specified for marking another path. It introduces the same effect as issueing `markStmtAsUsed(!SkipPath, S)`.

`llvm-cov` discovers counters in `FalseCount` to allocate `MaxCounterID` for empty profile data.

https://discourse.llvm.org/t/rfc-integrating-singlebytecoverage-with-branch-coverage/82492

Depends on: #112698 #112702 #112724

chapuni committed:
[Coverage] Make additional counters available for BranchRegion. NFC.

`getBranchCounterPair()` allocates an additional Counter to SkipPath
in `SingleByteCoverage`.

`IsCounterEqual()` calculates the comparison with rewinding counter
replacements.

`NumRegionCounters` is updated to take additional counters in account.

`incrementProfileCounter()` has a few additiona arguments.

- `UseSkipPath=true`, to specify setting counters for SkipPath. It
  assumes `UseSkipPath=false` is used together.

- `UseBoth` may be specified for marking another path. It introduces
  the same effect as issueing `markStmtAsUsed(!SkipPath, S)`.

`llvm-cov` discovers counters in `FalseCount` to allocate
`MaxCounterID` for empty profile data.

chapuni committed:
`first` may be cancelled.

Currently `first` is not None by default.

chapuni committed:
Merge branch 'users/chapuni/cov/single/nextcount-base' into users/chapuni/cov/single/nextcount

chapuni committed:
Update getSwitchImplicitDefaultCounter

chapuni committed:
Merge branch 'users/chapuni/cov/single/nextcount-base' into HEAD

chapuni committed:
Don't allocate second if SkipExpr isn't Expr.

ornata commented on clang/lib/CodeGen/CodeGenPGO.cpp:
```diff
@@ -1193,11 +1206,26 @@ std::pair<bool, bool> CodeGenPGO::getIsCounterPair(const Stmt *S) const {
 
-  unsigned Counter = (*RegionCounterMap)[S].first;
+  unsigned Counter;
+  auto &TheMap = (*RegionCounterMap)[S];
```
static_cast?

chapuni replied:
Do you mention the use of `auto`?

ornata commented on clang/lib/CodeGen/CodeGenPGO.cpp:
```diff
@@ -1193,11 +1206,26 @@ std::pair<bool, bool> CodeGenPGO::getIsCounterPair(const Stmt *S) const {
+  if (!UseSkipPath) {
+    if (!IsCounter.first)
+      return;
+    Counter = (TheMap.first & CounterPair::Mask);
```
This patch introduces a lot of code like

```
(X & CounterPair::Mask)
```

Would it make sense to create a small helper function to do this?

chapuni replied:
I encapsulated it. See #112724.

ornata commented on clang/lib/CodeGen/CodeGenPGO.cpp:
```diff
@@ -1193,11 +1206,26 @@ std::pair<bool, bool> CodeGenPGO::getIsCounterPair(const Stmt *S) const {
+    Counter = (TheMap.second & CounterPair::Mask);
+  }
+
+  if (!Builder.GetInsertBlock())
```
When can this happen? Comment?

chapuni replied:
It was the result of splitting the condition.
```
if (!RegionCounterMap || !Builder.GetInsertBlock())
    return;
```
to
```
if (!RegionCounterMap)
    return;

auto &TheMap = (*RegionCounterMap)[S]; // S should be allocated (as an initial value)
...

if (!Builder.GetInsertBlock())
    return;
```

ornata commented on clang/lib/CodeGen/CodeGenPGO.h:
```diff
@@ -112,6 +112,7 @@ class CodeGenPGO {
 public:
   std::pair<bool, bool> getIsCounterPair(const Stmt *S) const;
```
I think this function could use some documentation, or a better name.

This can happen in a later commit.

chapuni replied:
I couldn't name better.
```
return {I->second.Executed.hasValue(), I->second.Skipped.hasValue()};
```
It is defined in #112724.

ornata commented on clang/lib/CodeGen/CoverageMappingGen.cpp:
```diff
@@ -887,6 +887,9 @@ struct CounterCoverageMappingBuilder
   llvm::DenseMap<const Stmt *, CounterPair> &CounterMap;
 
+  CounterExpressionBuilder::ReplaceMap MapToExpand;
+  unsigned NextCounterNum;
```
Explicitly initialize to 0?

chapuni replied:
The only user shall initialize it.

ornata replied:
OK

ornata commented on clang/lib/CodeGen/CoverageMappingGen.cpp:
```diff
@@ -943,19 +942,43 @@ struct CounterCoverageMappingBuilder
     if (OutCount == ParentCount)
       return true;
 
+    // Try comaparison with pre-replaced expressions.
```
Can you put a small example in the comment here?

ornata commented on llvm/lib/ProfileData/Coverage/CoverageMapping.cpp:
```diff
@@ -638,6 +638,10 @@ static unsigned getMaxCounterID(const CounterMappingContext &Ctx,
   unsigned MaxCounterID = 0;
   for (const auto &Region : Record.MappingRegions) {
     MaxCounterID = std::max(MaxCounterID, Ctx.getMaxCounterID(Region.Count));
+    if (Region.Kind == CounterMappingRegion::BranchRegion ||
```
I feel like it would be nice to add helpers for Region.Kind just to make the code a little easier to read.

E.g.

```
isBranchRegion(Region) { return Region.Kind == CounterMappingRegion::BranchRegion; } 
isMCDCBranchRegion(Region)
```
Maybe in a follow-up commit?

chapuni replied:
I have 5a5838fba37153adb7885c897131dda09227eb2d

ornata replied:
OK cool

ornata commented on llvm/lib/ProfileData/Coverage/CoverageMapping.cpp:
```diff
@@ -638,6 +638,10 @@ static unsigned getMaxCounterID(const CounterMappingContext &Ctx,
     MaxCounterID = std::max(MaxCounterID, Ctx.getMaxCounterID(Region.Count));
+    if (Region.Kind == CounterMappingRegion::BranchRegion ||
+        Region.Kind == CounterMappingRegion::MCDCBranchRegion)
+      MaxCounterID =
```
Can we have a `updateMaxCounterID` helper? I think this happens in a couple places in the code now.

chapuni replied:
I guess there was no idea in initial designs to increase `Counter`s dynamically. It shall be improved in the future.

chapuni committed:
Merge branch 'users/chapuni/cov/single/nextcount-base' into users/chapuni/cov/single/nextcount

Conflicts:
	clang/lib/CodeGen/CoverageMappingGen.cpp

chapuni committed:
Fold either in switchcase

chapuni committed:
Introduce CounterMappingRegion::isBranch(). NFC.

Conflicts:
	clang/lib/CodeGen/CoverageMappingGen.cpp
	llvm/include/llvm/ProfileData/Coverage/CoverageMapping.h

chapuni commented:
I've accidentally closed this with pushing the wrong base. Please move to the re-created #120930 .
